modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -520,18 +518,84 @@ private CompletableFuture<HybridTimestamp> validateTableExistence(ReplicaRequest
         }
 
         return schemaSyncService.waitForMetadataCompleteness(opStartTs)
-                .thenApply(unused -> {
-                    schemaCompatValidator.failIfTableDoesNotExistAt(opStartTs, tableId());
+                .thenRun(() -> schemaCompatValidator.failIfTableDoesNotExistAt(opStartTs, tableId()));
+    }
+
+    /**
+     * Makes sure that {@link SchemaVersionAwareReplicaRequest#schemaVersion()} sent in a request matches table schema version
+     * corresponding to the operation.
+     *
+     * @param request Replica request corresponding to the operation.
+     * @param opTsIfDirectRo Operation timestamp for a direct RO, {@code null} otherwise.
+     * @return Future completed when the validation is finished.
+     */
+    private CompletableFuture<Void> validateSchemaMatch(ReplicaRequest request, @Nullable HybridTimestamp opTsIfDirectRo) {
+        if (!(request instanceof SchemaVersionAwareReplicaRequest)) {
+            return completedFuture(null);
+        }
+
+        SchemaVersionAwareReplicaRequest versionAwareRequest = (SchemaVersionAwareReplicaRequest) request;
+
+        HybridTimestamp tsToWaitForSchema = getTxStartTimestamp(request);
```

Comment:
It may be a good idea to pass `opTsIfDirectRo` to `getTxStartTimestamp` so that it will return it instead of `null`, this will make the code shorter

Reply:
Well, `getTxStartTimestamp()` returns the timestamp corresponding to the start of a transaction. I'm not sure that we should stretch the direct RO operation timestamp to think about it as a transaction start ts.

Reply:
as you wish